File: articles/al-kindi-frequency-analysis-and-the-birth-of-cryptanalysis/haskell/FrequencyCount.hs
module Main where

import Data.Char (isAlpha, toLower)
import qualified Data.Map.Strict as M

main :: IO ()
main = do
  let text = "THIS IS A TOY CLASSICAL CIPHER EXAMPLE"
      symbols = map toLower (filter isAlpha text)
      counts = foldr (\c -> M.insertWith (+) c 1) M.empty symbols
  print counts

File: articles/al-kindi-frequency-analysis-and-the-birth-of-cryptanalysis/rust/frequency_count.rs
use std::collections::HashMap;

fn main() {
    let text = "THIS IS A TOY CLASSICAL CIPHER EXAMPLE";
    let mut counts: HashMap<char, usize> = HashMap::new();
    for ch in text.to_lowercase().chars().filter(|c| c.is_alphabetic()) {
        *counts.entry(ch).or_insert(0) += 1;
    }
    println!("{:?}", counts);
}
